node/engine/engine.go

```diff
@@ -547,6 +595,64 @@ func (e *Engine) handlePaymentRequest(request PaymentRequest) (EngineEvent, erro
 	return ee, e.executeSideEffects(se)
 }
 
+// handleUpdateChannelRequest handles an UpdateChannelRequest (triggered by a client API call).
+// It creates new state of channel, signs it and sends it to the counterparty.
+func (e *Engine) handleUpdateChannelRequest(request UpdateChannelRequest) (EngineEvent, error) {
+	ee := EngineEvent{}
+	cId := request.ChannelId
+	c, ok := e.store.GetChannelById(cId)
+	if !ok {
+		return ee, fmt.Errorf("handleAPIEvent: Could not get channel from the store %s", cId)
+	}
+
+	s := directs.GetLatestProposedStateByAppData(c, request.AppData)
+
+	// If there is no suitable proposed state, creates a new one
+	if s == nil {
+		latestSignedState, err := c.LatestSignedState()
+		if err != nil {
+			return ee, fmt.Errorf("handleAPIEvent: Could not get latest supported state from the channel %s", cId)
+		}
+
+		newState := latestSignedState.Clone().State()
+		newState.AppData = request.AppData
+		newState.TurnNum = newState.TurnNum + 1
+
+		s = &newState
+	}
+
+	ss, err := c.SignAndAddState(*s, e.store.GetChannelSecretKey())
```

Comment:
Here we are countersigning a state if the appdata matched what came over the API.

We need to be careful: the state may have a different [**Outcome**](https://docs.statechannels.org/protocol-tutorial/0030-outcomes/). I suggest we need to add the outcome to the relevant data structures and check against that. 

A more elegant solution would be to use the [**state hash**](https://docs.statechannels.org/protocol-tutorial/0010-states-channels/#state-commitments) to check if it is "the same state".